File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        dists = [[10001 if i != j else 0 for i in range(n)] for j in range(n)]
        
        for fr, t, w in edges:
            dists[fr][t] = min(dists[fr][t], w)
            dists[t][fr] = min(dists[t][fr], w)
        
        flag = False
        for _ in range(n):
            for r in range(n):
                for c in range(n):
                    if r == c: #원점일 경우 넘어가기
                        continue
                    for tr in range(n):
                        if dists[r][c] > dists[r][tr] + dists[tr][c]:
                            flag = True
                            dists[r][c] = dists[r][tr] + dists[tr][c]
            if not flag: #더이상 바뀌는 값이 없을 경우
                break
            flag = False
            
        answer_list = [sum(d <= distanceThreshold for d in dist) for dist in dists]
        return max([i for i, val in enumerate(answer_list) if val == min(answer_list)])

File: 1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/1334-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] dists = new int[n][n];
        int[] dCnt = new int[n];
        List<Integer> answerList = new ArrayList<>();
        int minDist = 101;
        
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                dists[i][j] = 10001;
            }
            dists[i][i] = 0;
        }
        int from, to, weight;
        
        for(int[] edge : edges) {
            from = edge[0]; to = edge[1]; weight = edge[2];
            dists[from][to] = Math.min(dists[from][to], weight);
            dists[to][from] = Math.min(dists[to][from], weight);
        }
        
        // System.out.println(Arrays.deepToString(dists).replace("],", "], \n"));
        
        for(int x = 0; x < n; x++) {
            for(int i = 0; i < n; i++) {
                for(int j = 0; j < n; j++) {
                    if(i == j) continue;
                    for(int k = 0; k < n; k++) {
                        dists[i][j] = Math.min(dists[i][j], dists[i][k] + dists[k][j]);
                    }
                    if(x == n - 1) { //마지막 순회일 때
                        if(dists[i][j] <= distanceThreshold) {
                            dCnt[i]++;
                        }
                    }
                }
                if(x == n - 1) { //마지막 순회일 때
                    if(dCnt[i] < minDist) {
                        answerList.clear();
                        minDist = dCnt[i];
                        answerList.add(i);
                    } else if(dCnt[i] == minDist) {
                        answerList.add(i);
                    }
                }
            }
        }
        
        // System.out.println("\n" + Arrays.deepToString(dists).replace("],", "], \n"));
        // System.out.println("Dist Count: " + Arrays.toString(dCnt));
        // System.out.println("Answer List: " + answerList);
        
        Collections.sort(answerList, Collections.reverseOrder());
        return answerList.get(0);
    }
}